src/util/parented_ptr.h

```diff
@@ -2,83 +2,93 @@
 #define UTIL_PARENTED_PTR_H
 
 #include <QPointer>
+
 #include "util/assert.h"
+#include "util/memory.h"
 
 /**
  * Use this wrapper class to clearly represent a raw pointer that is owned by the QT object tree.
  * Objects which both derive from QObject AND have a parent object, have their lifetime governed by the QT object tree, 
  * and thus such pointers do not require a manual delete to free the heap memory when they go out of scope.
 **/
-template <typename T>
+template<typename T>
 class parented_ptr {
   public:
-    explicit parented_ptr(T* t) : m_pObject(t) {
-        DEBUG_ASSERT(t->parent() != nullptr);
-    }
+    parented_ptr() noexcept = default;
 
-    /* If U* is convertible to T* then we also want parented_ptr<U> convertible to parented_ptr<T> */
-    template <typename U>
-    parented_ptr(parented_ptr<U>&& u, typename std::enable_if<std::is_convertible<U*, T*>::value, void>::type * = 0)
-            : m_pObject(u.get()) {
-        DEBUG_ASSERT(u->parent() != nullptr);
+    parented_ptr(nullptr_t) noexcept
+            : m_ptr{nullptr} {
     }
 
-#if defined(__GNUC__) && __GNUC__ < 5
-    // gcc 4.8 does not implicit use the typ conversion move constructor
-    // from above when returning form a function and use finally RVO.
-    // It requires
-    //return std::move(pObject)
-    // This hack avoids it:
-    template <typename U>
-    operator parented_ptr<U>() const {
-        static_assert(std::is_convertible<T*, U*>::value,
-                "No implicit conversion from T* to U* found.");
-        return parented_ptr<U>(this->get());
+    explicit parented_ptr(T* t) noexcept
+            : m_ptr{t} {
     }
 
-#endif
+    ~parented_ptr() noexcept {
+        if (m_ptr != nullptr) {
+            DEBUG_ASSERT(m_ptr->parent() != nullptr);
+        }
+    }
 
     // Delete copy constructor and copy assignment operator
     parented_ptr(const parented_ptr<T>&) = delete;
     parented_ptr& operator=(const parented_ptr<T>&) = delete;
 
-    parented_ptr() : m_pObject(nullptr) {}
+    /*
+     * If U* is convertible to T* then we also want parented_ptr<U> convertible to parented_ptr<T>
+     */
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    parented_ptr(parented_ptr<U>&& u) noexcept
+            : m_ptr{u.m_ptr} {
+        u.m_ptr = nullptr;
+    }
+
+    /*
+     * If U* is convertible to T* then we also want parented_ptr<U> assignable to parented_ptr<T>
+     * E.g. parented_ptr<Base> base = make_parented<Derived>(); should work as expected.
+     */
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    parented_ptr& operator=(parented_ptr<U>&& u) noexcept {
+        m_ptr = u.m_ptr;
```

Comment:
m_ptr should be handled the same way as in the destructor, i.e. we should apply the same debug assertion as in the destructor before abandoning it. This could be wrapped into a function with a concise name for documentation and to avoid code duplication.

Reply:
You are totally right. That's a bug I made.

Reply:
Instead of using the extra function, I used the copy-swap idiom (in this case move-swap) in my commit I just pushed.